### ISSUE TITLE
Hang when calling method from extended C++ class in future

### ISSUE THREAD

**Issue by ibell** [pybind/pybind11#813]
I am working on extending a concrete (but derived) C++ class in python, and then calling the overridden function in a C++ thread (via an async).  I call the function ``added_fcn`` in the C++ level, which calls the python implementation, and then it just gets stuck (hangs) and never returns a value.  In the debugger, I can see it is in the wait function, and never leaves.

For what it's worth, originally I had hoped to just pass a std::function to my class, but it suffers from the same hanging behavior as extending a class.  Calling the function, or the overloaded method (not from a thread), works totally fine.

OSX, Xcode, python 2.7

#### Reproducible example code

interface_pybind11.cxx
``` c++
#include <pybind11/pybind11.h>
#include <pybind11/stl.h>
#include <pybind11/operators.h>
#include <pybind11/functional.h>

namespace py = pybind11;

class A{
public:
    virtual void pure_virt() = 0;
};

class B : public A{
public:
    void pure_virt() override {

    };
    virtual int added_fcn(){
        return -1;
    }
};

class PyB : public B {
    public:
        /* Inherit the constructors */
        using B::B;

        /* Trampoline (need one for each virtual function) */
        int added_fcn() override {
            PYBIND11_OVERLOAD(
                int,                      /* Return type */
                B,            /* Parent class */
                added_fcn,                   /* Name of function in C++ (must match Python name) */
                                            /* Argument(s) */
            );
        }
};

void call_parallel(const std::shared_ptr<B> &BB) {
    // This function hangs
    std::future<void> f2 = std::async(std::launch::async, [BB](){ BB->added_fcn() ; }); 
    f2.wait();
}

void init_fitter(py::module &m){

    py::class_<A>(m, "A");

    py::class_<B, PyB, A, std::shared_ptr<B> >(m, "B")
        .def(py::init<>())
        .def("added_fcn", &B::added_fcn);

    m.def("call_parallel", call_parallel);
}

PYBIND11_PLUGIN(PolyFitter) {
    py::module m("PolyFitter", "PolyFitter module");
    init_fitter(m);
    return m.ptr();
}
```

test.py
``` python
    import sys; sys.path.append('build/pybind11/Debug')
    import PolyFitter as pf
    class B(pf.B):
        def __init__(self):
            super(B, self).__init__()
            pass

        def added_fcn(self):
            return 3

    b = B()
    pf.call_parallel(b)
```

**Comment by dean0x7d** [comment]
You need to release Python's GIL before calling back to Python from C++ threads. See the [Global Interpreter Lock](http://pybind11.readthedocs.io/en/master/advanced/misc.html#global-interpreter-lock-gil) in the docs.

Essentially, `call_parallel` needs `py::gil_scoped_release`. The `py::gil_scoped_acquire` part is usually taken care of automatically, e.g. the `PYBIND11_OVERLOAD` macro and the `std::function` casters always acquire the GIL before calling Python functions. It's just the GIL release part that needs to be done explicitly when threads are spawned in C++.

**Comment by ibell** [comment]
Calling a function directly with a future seems to work.  This is fine:
``` c++
m.def("call_f", [](const std::function<int(int)> &f, int x) -> int {
        /* Release GIL before calling into (potentially long-running) C++ code */
        py::gil_scoped_release release;
        std::future<int> fv = std::async(std::launch::async, [f, x]() -> int{return f(x);} );
        return fv.get();
    });
```
But calling a python method of a class derived in python still hangs.  I get this mysterious message in my console when I start up my application (at import time?):
```
objc[90449]: Class _NSZombie_OS_xpc_array is implemented in both ?? and ??. One of the two will be used. Which one is undefined.
```
Perhaps that is related?

**Comment by ibell** [comment]
For those that end up here, here is a layout of what the GIL stuff should look like:
``` C++
#include <pybind11/pybind11.h>
#include <pybind11/stl.h>
#include <pybind11/operators.h>
#include <pybind11/functional.h>

namespace py = pybind11;

class A{
public:
    virtual void pure_virt() = 0;
};

class B : public A{
public:
    void pure_virt() override {

    };
    virtual int added_fcn(){
        return -1;
    }
};

class PyB : public B {
    public:
        /* Inherit the constructors */
        using B::B;

        /* Trampoline (need one for each virtual function) */
        int added_fcn() override {
            
            /* Acquire GIL before calling Python code */
            py::gil_scoped_acquire acquire;

            PYBIND11_OVERLOAD(
                int,               /* Return type */
                B,                 /* Parent class */
                added_fcn,         /* Name of function in C++ (must match Python name) */
                                   /* Argument(s) */
            );
        }
};

void call_parallel(const std::shared_ptr<B> &BB) {
    std::future<void> f2 = std::async(std::launch::async, [BB](){ BB->added_fcn() ; }); 
    f2.get();
}

void init_fitter(py::module &m){

    py::class_<A, std::shared_ptr<A>>(m, "A");

    py::class_<B, PyB, A, std::shared_ptr<B> >(m, "B")
        .def(py::init<>())
        .def("added_fcn", &B::added_fcn);

    m.def("call_parallel", call_parallel, py::call_guard<py::gil_scoped_release>());
}

PYBIND11_PLUGIN(PolyFitter) {
    py::module m("PolyFitter", "PolyFitter module");
    init_fitter(m);
    return m.ptr();
}
```

**Comment by ibell** [comment]
I think I might not understand the GIL, perhaps that is the issue here.  

I have multiple threads, each of which owns a (a few) class instances in C++.  If each class instance is pure C++ (without calling to python), all is well.  Now, in my algorithm, I evaluate a function for all threads in parallel (that's the point).  If all class instances don't break the C++|python barrier, its all still fine since the calculations are all happening at the C++ level.

As soon as I start calling back into python, I start running into trouble.  I can't modify the C++ itself, thus only the trampoline class overload is allowed to play with the GIL.  And I have multiple C++ threads calling back into python in parallel.  Is this a pattern that is possible?  

Basically, I want to be able to extend C++ classes with python (because writing python is much easier than writing C++), while retaining the power of my algorithm in C++.

**Comment by ibell** [comment]
To solve my particular problem, I had to, in the trampoline, release and then acquire the GIL, something like:

``` C++
class PyFiniteDiffOutput : public FiniteDiffOutput {
    public:
        /* Inherit the constructors */
        using FiniteDiffOutput::FiniteDiffOutput;

        /* Trampoline (need one for each virtual function) */
        double call_func(const std::vector<double> &c) override {
            /* Release the GIL */
            py::gil_scoped_release release;
            {
                /* Acquire GIL before calling Python code */
                py::gil_scoped_acquire acquire;
                
                PYBIND11_OVERLOAD(
                    double,                      /* Return type */
                    FiniteDiffOutput,            /* Parent class */
                    call_func,                   /* Name of function in C++ (must match Python name) */
                    c                            /* Argument(s) */
                );
            }
        }
    };
```

That still leaves unresolved what to do about calling python functions in a number of C++ threads in parallel.  Is that possible?

**Comment by RenegadeTech** [comment]
I had a similar problem, what I did was change ```std::launch::async``` to ```std::launch::deferred``` and it works fine.
Not 100% sure why though...
